server/application/application.go

```diff
@@ -498,134 +383,60 @@ func (s *Server) getRepo(ctx context.Context, repoURL string) *appv1.Repository
-			err = json.Unmarshal([]byte(resourceState.LiveState), liveObj)
-			if err != nil {
-				return nil, nil, err
-			}
-		}
+func (s *Server) Rollback(ctx context.Context, rollbackReq *ApplicationRollbackRequest) (*appv1.Application, error) {
+	return s.setAppOperation(ctx, rollbackReq.Name, func(app *appv1.Application) (*appv1.Operation, error) {
 
-		if resourceState.TargetState != "null" {
-			targetObj = &unstructured.Unstructured{}
-			err = json.Unmarshal([]byte(resourceState.TargetState), targetObj)
-			if err != nil {
-				return nil, nil, err
+		var deploymentInfo *appv1.DeploymentInfo
+		for _, info := range app.Status.RecentDeployments {
+			if info.ID == rollbackReq.ID {
+				deploymentInfo = &info
+				break
 			}
 		}
-
-		needsCreate := liveObj == nil
-		needsDelete := targetObj == nil
-
-		obj := targetObj
-		if obj == nil {
-			obj = liveObj
-		}
-		resDetails := ResourceDetails{
-			Name:      obj.GetName(),
-			Kind:      obj.GetKind(),
-			Namespace: namespace,
+		if deploymentInfo == nil {
+			return nil, status.Errorf(codes.InvalidArgument, "application %s does not have deployment with id %v", rollbackReq.Name, rollbackReq.ID)
 		}
 
-		if resourceState.Status == appv1.ComparisonStatusSynced {
-			resDetails.Message = fmt.Sprintf("already synced")
-		} else if dryRun {
-			if needsCreate {
-				resDetails.Message = fmt.Sprintf("will create")
-			} else if needsDelete {
-				if prune {
-					resDetails.Message = fmt.Sprintf("will delete")
-				} else {
-					resDetails.Message = fmt.Sprintf("will be ignored (should be deleted)")
-				}
-			} else {
-				resDetails.Message = fmt.Sprintf("will update")
-			}
-		} else {
-			if needsDelete {
-				if prune {
-					err = kube.DeleteResource(config, liveObj, namespace)
-					if err != nil {
-						return nil, nil, err
-					}
+		return &appv1.Operation{
+			Sync: &appv1.SyncOperation{
+				NoOverrides: false,
+				Overrides:   deploymentInfo.ComponentParameterOverrides,
+				Revision:    deploymentInfo.Revision,
+				Prune:       rollbackReq.Prune,
+				DryRun:      rollbackReq.DryRun,
+			},
+		}, nil
+	})
+}
 
-					resDetails.Message = fmt.Sprintf("deleted")
-				} else {
-					resDetails.Message = fmt.Sprintf("ignored (should be deleted)")
-				}
-			} else {
-				_, err := kube.ApplyResource(config, targetObj, namespace)
-				if err != nil {
-					return nil, nil, err
-				}
-				if needsCreate {
-					resDetails.Message = fmt.Sprintf("created")
-				} else {
-					resDetails.Message = fmt.Sprintf("updated")
-				}
-			}
-		}
-		syncRes.Resources = append(syncRes.Resources, &resDetails)
+func (s *Server) setAppOperation(ctx context.Context, appName string, operationCreator func(app *appv1.Application) (*appv1.Operation, error)) (*appv1.Application, error) {
+	app, err := s.Get(ctx, &ApplicationQuery{Name: appName})
+	if err != nil {
+		return nil, err
+	}
+	if app.Operation != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "other operation is already in progress")
```

Comment:
lets use the message: "another operation is already in progress".

One thing to think about is possibly making this idempotent. If the operation is identical, then we can return success. Not sure if this is truly correct behavior though.

Reply:
Updated message.

If app resources were deleted manually then user want to sync even if the previous sync operation is identical and successful.